src/runtime/thread_pool.cc

```diff
@@ -127,99 +130,138 @@ class ParallelLauncher {
-    } else {
-      size_t old_size = ring_.size();
-      ring_.resize(old_size * 2);
-      if (head_ + num_pending_ > old_size) {
-        // copy the ring overflow part into the tail.
-        size_t ncopy = head_ + num_pending_ - old_size;
-        memcpy(&ring_[0] + old_size, &ring_[0], ncopy * sizeof(Task));
-      }
-      ring_[(head_ + num_pending_) % ring_.size()] = task;
-      ++num_pending_;
+  void Push(const Task& input) {
+    while (!Enqueue(input)) {
+      std::this_thread::yield();
     }
-    if (nwait_consumer_ != 0) {
-      lock.unlock();
+    if (pending_.fetch_add(1) == -1) {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.notify_one();
     }
   }
+
   /*!
-   * \brief Pop task from the queue
-   * \param task The task to be poped.
-   * \param timeout The number of cycles to spin before sleep.
-   * \return Whether pop is successful or we need to exit now.
+   * \brief Pop a task out of the queue and condition wait if no tasks.
+   * \param output The pointer to the task to be dequeued.
+   * \param timeout The number of iterations to spin before sleep.
+   * \return Whether pop is successful (true) or we need to exit now (false).
    */
-  bool Pop(Task* task, int timeout = 10) {
-    std::unique_lock<std::mutex> lock(mutex_);
-    if (num_pending_ != 0) {
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
-    } else {
-      lock.unlock();
-      // do a bit spin and busy waiting before sleep.
-      for (int i = 0; i < timeout && num_pending_ == 0; ++i) {
+  bool Pop(Task* output, uint32_t timeout = 100000) {
+    // busy wait a bit when the queue is empty
+    // if a new task comes to the queue quickly, this wait avoid the worker from sleeping
+    for (uint32_t i = 0; i < timeout && pending_.load() == 0; ++i) {
         std::this_thread::yield();
       }
-      lock.lock();
-      ++nwait_consumer_;
+    if (pending_.fetch_sub(1) == 0) {
+      std::unique_lock<std::mutex> lock(mutex_);
       cv_.wait(lock, [this] {
-          return num_pending_ != 0 || exit_now_.load();
+          return pending_.load() >= 0 || exit_now_.load();
         });
-      --nwait_consumer_;
-      *task = ring_[head_];
-      head_ = (head_ + 1) % ring_.size();
-      --num_pending_;
-      if (exit_now_.load()) return false;
     }
-    return true;
+    return Dequeue(output);
+    }
+
+  /*!
+   * \brief Signal to terminate the worker.
+   */
+  void SignalForKill() {
+    exit_now_.store(true);
+    cv_.notify_all();
   }
 
  private:
-  // Number of the elments in the queue
-  uint32_t num_pending_{0};
-  // Queue head
-  uint32_t head_{0};
-  // Number of consumers to wait.
-  uint32_t nwait_consumer_{0};
+  struct node_t {
+    node_t* next;
+    Task data;
+  };
+
+  typedef std::aligned_storage<sizeof(Task), std::alignment_of<Task>::value>::type \
+    aligned_t;
+
+  typedef char cache_line_pad_t[64];
```

Comment:
add comment here why we need to do this.

define a constant kL1CacheBytes in the beginning of this file, we will likely need to use macro to switch the size. For ARM it can be 32

Reply:
I will leave it as 64 for now. We can add macro switch later if needed.